src/wwwroot/app/directives/d3-histogram/d3-histogram.js

```diff
@@ -0,0 +1,357 @@
+              }
+           });
+           p.style('fill', (element) => {
+              return color(element.name);
+           });
+        });
+
+      legend.append('text')
+        .attr('x', width - 24)
+        .attr('y', 9)
+        .attr('dy', '.35em')
+        .style('text-anchor', 'end')
+        .text((d) => {
+           return d;
+        });
+   }
+
+   function generatePie(location, pieData) {
+      let pieKeys = map(pieData, (d) => {
+         return d.name;
+      });
+      let sumValues = sumBy(pieData, (d) => {
+         return d.number;
+      });
+
+      let width = 680,
+         height = 450,
+         radius = Math.min(width, height) / 2;
+
+      let color = d3.scaleOrdinal()
+         .range(elementColors);
+
+      let arc = d3.arc()
+         .outerRadius(radius - 10)
+         .innerRadius(30);
+
+      let labelArc = d3.arc()
+         .outerRadius(radius - 40)
+         .innerRadius(radius - 40);
+
+      let pie = d3.pie()
+         .sort(null)
+         .value((d) => {
+            return d.number;
+         });
+      let arcs = pie(pieData);
+
+      let svg = d3.select('.pie-container').append('svg')
+         .attr('class', 'pie-svg-cont')
+         .attr('width', width)
+         .attr('height', height)
+         .append('g')
+         .attr('transform', `translate( ${width - 420}, ${height / 2} )`);
+
+      let g = svg.selectAll('.arc')
+         .data(arcs)
+         .enter()
+         .append('g')
+         .attr('class', 'arc');
+
+      g.append('path')
+      .attr('d', arc)
+      .style('fill', (d) => {
+         return color(d.index);
+      });
+
+      g.append('text')
+      .attr('transform', (d) => {
+         return `translate( ${labelArc.centroid(d)} )`;
+      })
+      .attr('class', 'pie-piece-value')
+      .attr('dy', '.35em')
+      .text((d) => {
+         return `${round((d.value / sumValues) * 100)}%`;
+      });
+
+      let legend = svg.selectAll('.legend')
+        .data(pieKeys)
+        .enter().append('g')
+        .attr('class', 'legend')
+        .attr('transform', (d, i) => {
+           return `translate( 160, ${(i * 20) - 200} )`;
+        });
+
+      legend.append('rect')
+        .attr('x', width / 3)
+        .attr('y', 0)
+        .attr('width', 18)
+        .attr('height', 18)
+        .style('fill', (element, i) => {
+           return color(i);
+        });
+
+      legend.append('text')
+        .attr('class', 'pie-legend-text')
+        .attr('x', (width / 3) - 18)
+        .attr('y', 9)
+        .attr('dy', '.35em')
+        .style('text-anchor', 'end')
+        .text((d, i) => {
+           return `${d}: ${pieData[i].number}`;
+        });
+      //open modal window afrer creating svg ekement with pie chart
+      openModalWindowForPieChart(location);
+      return svg;
+   }
+
+   function openModalWindowForPieChart(location) {
+      let pieElement = document.getElementById('pie-cont').innerHTML;
+      let onCloseModalElement =  () =>  {
```

Comment:
you can move this non-named function to object definiton, instead of creating a variable